Fix installation error.

IonicaBizau opened the pull request:
Fixes #13997. Check if the `bin` object exists to prevent the errors.

IonicaBizau committed:
Fix installation error.

Fixes #13997. Check if the `bin` object exists to prevent the errors.

IonicaBizau commented:
Not quite sure why the tests are failing on Windows, but I guess it's not because my change.

watilde commented on lib/install/deps.js:
```diff
@@ -600,6 +600,7 @@ var earliestInstallable = exports.earliestInstallable = function (requiredBy, tr
   var binaryMatches = typeof pkg.bin === 'object' && tree.children.some(function (child) {
     if (child.removed) return false
     if (typeof child.package.bin !== 'object') return false
+    if (!pkg.bin || !child.package.bin) return false
```
Thanks for the good catch/patch! I think the actual point of the bug is the conditional expression `typeof variable === 'object'` and `typeof variable !== 'object'` . It's sometimes annoy since the variable could be `null` at the case and `typeof null` is `object`. So checking `null` would be better imo, like the following:
```diff
-- var binaryMatches = typeof pkg.bin === 'object' && tree.children.some(function (child) {
++ var binaryMatches = pkg.bin !== null && typeof pkg.bin === 'object' && tree.children.some(function (child) {
     if (child.removed) return false
++   if (child.package.bin === null) return false
     if (typeof child.package.bin !== 'object') return false
       return Object.keys(child.package.bin).some(function (bin) {
         return pkg.bin[bin]
     })
   })
```
Cheers!

iarna replied:
The only values that are supposed to come out of `read-package-json`/`normalize-package-json` are an array or undefined/null. (If you had a string in your `package.json` that gets converted to an object before this code can see it.)

Truthyness is sufficient for `pkg.bin` but `child.package.bin` needs truthyness (or not null) checking AND object-ness-checking as it's gonna call `Object.keys`.

My inclination is to have the guard be the same for both fields.

IonicaBizau replied:
@iarna Makes sense. Will definitely fix that. I was actually thinking the same, but I thought there are other reasons to check against the type. 👍

IonicaBizau replied:
In my latest commit I changed `typeof pkg.bin === 'object' && ...` into `pkg.bin && ...`. 👍 

@iarna @watilde Can you review it and help me to improve the tests? 💭

watilde replied:
LGTM for the last change! Great :smiley_cat:

In this case, the package that has `bin: null` is [cli-size](https://github.com/IonicaBizau/cli-size/blob/master/package.json#L38) which is used by `cli-update`, so I've written the test case to reproduce the error:
https://gist.github.com/watilde/2e2533299bf2942bd719893062017307

Here is a small example to reproduce the original issue that I do some spelunking:
```json
{
  "name": "test",
  "version": "0.0.0",
  "dependencies": {
    "diable": "*",
    "cli-update": "*",
  }
}
```

Thanks!

iarna commented:
Thank you for putting this together. After reviewing the code in `read-package-json` and `normalize-package-json` I can confirm that a guard here is appropriate. The one thing that we'll need before I can land this is a test.

IonicaBizau commented:
> The one thing that we'll need before I can land this is a test.

@iarna Can one of you help me writing a test for this? Should we directly call the `findRequirement` method with data that was making the things crash before?

Where would be the appropriate place to put the test in?

Thanks! 🍀

IonicaBizau commented:
I added a test which is failing without this change and is passing with the patch.

But I feel I didn't do it the right way. Can you check that and help me improve it if it's the case? 😊

I also simplified the fix by simply checking `pkg.bin && ...` since `pkg.bin` can be `null` which is an object. ✨

watilde commented:
In this case, the package that has `bin: null` is [cli-size](https://github.com/IonicaBizau/cli-size/blob/master/package.json#L38) which is used by `cli-update`, so I've written the test case to reproduce the error:
https://gist.github.com/watilde/2e2533299bf2942bd719893062017307

Here is a small example `packagejson` to reproduce the original issue that I do some spelunking:
```json
{
  "name": "test",
  "version": "0.0.0",
  "dependencies": {
    "diable": "*",
    "cli-update": "*",
  }
}
```

Cheers!

IonicaBizau commented:
@watilde Thanks a lot for that! :cake: I just added it. :rocket:

IonicaBizau commented:
I'm not sure how the process works––I guess this is ready to be merged, or should I do anything else? :thought_balloon:

iarna commented:
Regarding process: This is the week of our next release, so I'm reviewing outstanding PRs. =)

iarna reviewed (approved):
This test seems good to me. You asked for feedback:

These days I typically use the `tacks` module to handle creating and destroying my fixture directories.

And to make using it easier, I further use `scripts/maketest` to generate the scaffolding for my test.

IonicaBizau commented:
Sounds amazing! ✨

iarna commented:
This has been merged into the `release-next` branch and will be in the next release later today.
